File: matlab/pendulum.m
classdef pendulum < handle
	properties
	dt
	m
    l
    b
    g
	x_limits
	num_points_x
	x_dot_limits
	num_points_x_dot
    num_states
    num_actions
	Q
	R
	x
	goal
	end
	
	methods 
		function p = pendulum(m, l, b, g, dt, x_limits, num_points_x, x_dot_limits, num_points_x_dot, num_points_u, Q, R, goal)
			p.dt = dt;
			p.m = m;
            p.l = l;
            p.b = b;
            p.g = g;
			p.x_limits = x_limits;
			p.x_dot_limits = x_dot_limits;
			p.num_points_x = num_points_x;
			p.num_points_x_dot = num_points_x_dot;
            p.num_states = num_points_x*num_points_x_dot;
            p.num_actions = num_points_u;
			p.Q = Q;
			p.R = R;
			p.goal = goal;
		end

		function [x, r, is_goal] = step(p, u)
			x = zeros(2,1);
            %  u = u/p.m;
			x(1) = p.x(1) + p.x(2)*p.dt;
			x(2) = p.x(2) + (u - p.m*p.g*p.l*sin(p.x(1)) - p.b*p.x(2))/(p.m*p.l^2)*p.dt;
			% Wrap around
			if x(1) > p.x_limits(2)
				x(1) = x(1) - p.x_limits(2);
			end
			if x(1) < p.x_limits(1)
				x(1) = x(1) + p.x_limits(2);
			end
			if x(2) > p.x_dot_limits(2)
				x(2) = p.x_dot_limits(2);
			end
			if x(2) < p.x_dot_limits(1)
				x(2) = p.x_dot_limits(1);
			end
			p.x = x;
			r = (p.R*u^2 + (p.x - p.goal)'*p.Q*(p.x - p.goal))*p.dt;

			% Check if goal is reached
			is_goal = false;
			dx = (p.x_limits(2)-p.x_limits(1))/(p.num_points_x - 1); 
    		dx_dot = (p.x_dot_limits(2)-p.x_dot_limits(1))/(p.num_points_x_dot - 1);
    		if (abs(p.goal(1) - p.x(1))<dx && abs(p.goal(2)-p.x(2))<dx_dot)
    			is_goal = true;
    		end
		end

		function x = reset(p)
			x = zeros(2,1);
			x(1) = p.x_limits(1) + (p.x_limits(2) - p.x_limits(1))*rand();
			x(2) = p.x_dot_limits(1) + (p.x_dot_limits(2) - p.x_dot_limits(1))*rand();
			p.x = x;
        end
        
        function [s_, is_goal] = dynamics(p, s, a)
            s_ = zeros(size(s));
            s_(:,1) = s(:,1) + s(:,2)*p.dt;
            s_(:,2) = s(:,2) + (a - p.m*p.l*p.g*sin(s(:,1)) - p.b*s(:,2))/(p.m*p.l^2)*p.dt;
            
            s_(s_(:,1) > p.x_limits(2),1) = s_(s_(:,1) > p.x_limits(2),1) - p.x_limits(2);
			s_(s_(:,1) < p.x_limits(1),1) = s_(s_(:,1) < p.x_limits(1),1) + p.x_limits(2);
			s_(s_(:,2) > p.x_dot_limits(2),2) = p.x_dot_limits(2);
			s_(s_(:,2) < p.x_dot_limits(1),2) = p.x_dot_limits(1);
            
            dx = (p.x_limits(2)-p.x_limits(1))/(p.num_points_x - 1); 
    		dx_dot = (p.x_dot_limits(2)-p.x_dot_limits(1))/(p.num_points_x_dot - 1);
            is_goal = (abs(s(:,1)-p.goal(1)) < 0.5*dx) & (abs(s(:,2)-p.goal(2)) < 0.5*dx_dot);
            s_(is_goal,:) = s(is_goal,:);
        end
        
        function r = cost(p, s_, a)
            r = (p.R*a.^2 + sum(((s_ - repmat(p.goal',size(s_,1),1))*p.Q).*(s_ - repmat(p.goal',size(s_,1),1)),2))*p.dt;
        end
	end
end

File: python/env/pendulum.py
import numpy as np

class pendulum:
    def __init__(self, m, l, b, g, dt, goal, x_limits, dx, x_dot_limits, dx_dot, Q, R):
        self.m = m
        self.l = l
        self.b = b
        self.g = g
        self.dt = dt
        assert type(goal)==type(np.array([])) and goal.shape[0]==2 and goal.shape[1]==1, "Check goal!"
        self.goal = goal
        self.x_limits = x_limits
        self.dx = dx
        self.dx_dot = dx_dot
        self.x_dot_limits = x_dot_limits
        self.Q = Q
        self.R = R
        _ = self.reset()

    def reset(self):
        self.x = np.array([\
                  [self.x_limits[0] + (self.x_limits[1] - self.x_limits[0])*np.random.rand()],\
                  [self.x_dot_limits[0] + (self.x_dot_limits[1] - self.x_limits[0])*np.random.rand()]\
                  ])

        return self.x

    def step(self, u):
        x = np.zeros((2,1))
        x[0,0] = self.x[0,0] + self.x[1,0]*self.dt
        x[1,0] = self.x[1,0] + (u \
                      - self.m*self.g*self.l*np.sin(self.x[0,0])\
                      - self.b*self.x[1,0])/(self.m*self.l**2)*dt

        if (x[0,0] > self.x_limits[1]):
            x[0,0] = x[0,0] - self.x_limits[1]
        elif (x[0,0] < self.x_limits[0]):
            x[0,0] = x[0,0] + self.x_limits[1]

        if (x[1,0] > self.x_dot_limits[1]):
            x[1,0] = self.x_dot_limits[1]
        elif (x[1,0] < self.x_dot_limits[1]):
            x[1,0] = self.x_dot_limits[0]
        self.x = x

        r = (self.R*u**2 + np.dot((x - self.goal).T, np.dot(self.Q, x - self.goal)))*self.dt
        
        is_terminal = False
        if ((np.abs(self.goal[0,0] - x[0,0]) < dx) and (np.abs(self.goal[1,0] - x[1,0]) < dx_dot)):
            is_terminal = True

        return x, r, is_terminal

    def dynamics(self, states, actions):

        assert states.shape[0]==2, "states.shape[0] != 2"
        assert states.shape[1]==actions.shape[1], "states.shape[1] != actions.shape[1]"
        
        states_ = np.zeros(states.shape)
        states_[0,:] = states[0,:] + states[1,:]*self.dt
        states_[1,:] = states[1,:] + (actions - self.m*self.g*self.l*np.sin(states[0,:])\
                           - self.b*states[1,:])/(self.m*self.l**2)*dt

        states_[0, states_[0,:] > self.x_limits[1]] -= self.x_limits[1]
        states_[0, states_[0,:] < self.x_limits[0]] += self.x_limits[1]
        states_[1, states_[1,:] > self.x_dot_limits[1]] = self.x_dot_limits[1]
        states_[1, states_[1,:] < self.x_dot_limits[0]] = self.x_dot_limits[0]

        diff = states_ - np.repeat(self.goal, states_.shape[1])
        rewards = self.dt*(np.power(diff[0,:],2)*self.Q[0,0] + np.power(diff[1,:],2)*self.Q[1,1] + (self.Q[0,1] + self.Q[1,0])*diff[0,:]*diff[1,:])
        are_terminal = np.logical_and(abs(states_[0,:] - goal[0,0]) < dx, abs(states_[1,:] - goal[1,0]) < dx_dot)
        
        return states_, rewards, are_terminal

